sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -75,11 +75,13 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]]
 
   override val origin: Origin = CurrentOrigin.get
 
-  /**
-   * A mutable map for holding auxiliary information of this tree node. It will be carried over
-   * when this node is copied via `makeCopy`, or transformed via `transformUp`/`transformDown`.
-   */
-  private val tags: mutable.Map[TreeNodeTag[_], Any] = mutable.Map.empty
+  private[this] var _tags: mutable.Map[TreeNodeTag[_], Any] = null
+  private def tags: mutable.Map[TreeNodeTag[_], Any] = {
```

Comment:
nit: can we also add a code comment on what we're doing in this PR? (so people scanning the code understand why we might be doing this strange setting of the map)